book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+
+We can expand the left hand side into the 2 limb format in order to obtain 2 intermediate sums
+
+$$
+\begin{aligned}
+s_{01} = x_{01} + f_{01}' \\
+s_2 = x_2 + f'_2 \\
+\end{aligned}
+$$
+
+where $x_{01}$ and $f'_{01}$ are defined like this
+
+$$
+\begin{aligned}
+x_{01} = x_0 + 2^{\ell} \cdot x_1 \\
+f'_{01} = f'_0 + 2^{\ell} \cdot f'_1 \\
+\end{aligned}
+$$
+
+and $x$ and $f'$ are defined like this
+
+$$
+\begin{aligned}
+x = x_{01} + 2^{2\ell} \cdot x_2 \\
+f' = f'_{01} + 2^{2\ell} \cdot f'_2 \\
+\end{aligned}
+$$
+
+Going back to our intermediate sums, the maximum bit length of sum $s_{01}$ is computed from the maximum bit lengths of $x_{01}$ and $f'_{01}$
+
+$$
+\underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{x_{01}} + \underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{f'_{01}} = 2^{2\ell+ 1} - 2,
+$$
+
+which means $s_{01}$ is at most $2\ell + 1$ bits long.
+
+Similarly, since $x_2$ and $f'_2$ are less than $2^{\ell}$, the max value of $s_2$ is
+
+$$
+(2^{\ell} - 1) + (2^{\ell} - 1) = 2^{\ell + 1} - 2,
+$$
+
+which means $s_2$ is at most $\ell + 1$ bits long.
+
+Thus, we must constrain
+
+$$
+s_{01} + 2^{2\ell} \cdot s_2 - x'_{01} - 2^{2\ell} \cdot x'_2 = 0 \mod 2^t.
+$$
+
+The accumulation of this into parts looks like this.
+
+```text
+0             L             2L            3L=t          4L
+|-------------|-------------|-------------|-------------|-------------|
+                            :
+|------------s01------------:-| 2L + 1
+                            : ↖w01
+                            |------s2-----:-| L + 1
+                            :               ↖w2
+                            :
+|------------x'01-----------|
+                            :
+                            |------x'2----|
+                            :
+\____________________________/
+             ≈ z01           \_____________/
+                                   ≈ z2
+```
+
+The two parts are computed with
+
+$$
+\begin{aligned}
+z_{01} &= s_{01} - x'_{01} \\
+z_2 &= s_2 - x'_2.
+\end{aligned}
+$$
+
+Therefore, there are two carry bits $w_{01}$ and $w_2$ such that
+
+$$
+\begin{aligned}
+z_{01} &= 2^{2\ell} \cdot w_{01} \\
+z_2 + w_{01} &= 2^{\ell} \cdot w_2
+\end{aligned}
+$$
+
+In this scheme $x'_{01}, x'_2, w_{01}$ and $w_2$ are witness data, whereas $s_{01}$ and $s_2$ are formed from a constrained computation of witness data $x_{01}, x_2$ and constraint system public parameter $f'$.  Note that due to carrying, witness $x'_{01}$ and $x'_2$ can be different than the values $s_{01}$ and $s_2$ computed from the limbs.
+
+Thus, each bound addition $x + f'$ requires the following witness data
+
+- $x_{01}, x_2$
+- $x'_{01}, x'_2$
+- $w_{01}, w_2$
+
+where $f'$ is baked into the constraint system.  The following constraints are needed
+
+- $2^{2\ell} \cdot w_{01} = s_{01} - x'_{01}$
+- $2^{\ell} \cdot w_2 = s_2 + w_{01} - x'_2$
+- $x'_{01} \in [0, 2^{2\ell})$
+- $x'_2 \in [0, 2^{\ell})$
+- $w_{01} \in [0, 2)$
+- $w_2 \in [0, 2)$
```

Comment:
If I remember correctly, I think we wanted to avoid this high carry

Reply:
Yes, we can avoid it.  And I wrote a proof by contradiction that we can ignore it for bound additions here https://github.com/o1-labs/proof-systems/blob/bb8f7bf3c70b87c48e6238d9be4b3feae90992ed/kimchi/src/tests/foreign_field_mul.rs#L244

However, I didn't want to implement it and remove the witness until you and @matthew have checked it.

Anyway, it looks like we have a completely new approach now, so I'm not sure this is needed at all.  Thus, I would propose to do it in a subsequent PR if we still need to do the bound addition like this, which I think we may not.

Also, there was another cool update where we can remove another witness element and another constraint for scaled $carry_11$ but again, we can do that in the PR for the new version